Expanded readme (#13)
graphcore-research/kg-topology-toolbox: README.md

<h1 align="center">
  KG Topology Toolbox
</h1>

![Continuous integration](https://github.com/graphcore-research/kg-topology-toolbox/actions/workflows/ci.yaml/badge.svg)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

> <b>KG Topology Toolbox:</b> is a Python package designed to efficiently extract common topological patterns from knowledge graphs, 

Reviewer: Was this add between the two rows of badges intentional? Not convinced by the positioning, and also it seems we are repeating a lot in the first three text paragraphs now.

Author: You are right - it wasn't really needed and I have removed it

For example, we can load a KG from a CSV file:

```python
import pandas as pd
df = pd.read_csv("path/to/kg.csv", columns=["h", "r", "t"])
```

This can then be used to instantiate a `KGTopologyToolbox` object:

```python
kgtt = KGTopologyToolbox(df)
```

### Computing Edge Topological Patterns

The `KGTopologyToolbox` object can be used to compute the topological properties of the KG. For example, to compute the edge patterns of the KG, we can use the [`edge_pattern_summary`](https://graphcore-research.github.io/kg-topology-toolbox/generated/kg_topology_toolbox.topology_toolbox.KGTopologyToolbox.html#kg_topology_toolbox.topology_toolbox.KGTopologyToolbox.edge_pattern_summary) method:

```python
edge_eps = kgtt.edge_pattern_summary()
```

> This will return a DataFrame with the edge patterns of the KG, where values have been computed for each edge contained with the graph.

Reviewer: ```suggestion
This will return a DataFrame with the edge patterns of the KG, where values have been computed for each edge contained within the graph.
```

> The values computed by the [`edge_pattern_summary`](https://graphcore-research.github.io/kg-topology-toolbox/generated/kg_topology_toolbox.topology_toolbox.KGTopologyToolbox.html#kg_topology_toolbox.topology_toolbox.KGTopologyToolbox.edge_pattern_summary) method include: loop, symmetric, inverse, inference, composition, number of triangles and other pattern metrics.

Reviewer: ```suggestion
The values computed by the [`edge_pattern_summary`](https://graphcore-research.github.io/kg-topology-toolbox/generated/kg_topology_toolbox.topology_toolbox.KGTopologyToolbox.html#kg_topology_toolbox.topology_toolbox.KGTopologyToolbox.edge_pattern_summary) method include edge pattern type (loop, symmetric, inverse, inference, composition), number of triangles insisting on the edge and other pattern metrics.
```

### Computing Edge Cardinality Patterns

Similarly, to compute the cardinality patterns of the KG, we can use the [`edge_degree_cardinality_summary`](https://graphcore-research.github.io/kg-topology-toolbox/generated/kg_topology_toolbox.topology_toolbox.KGTopologyToolbox.html#kg_topology_toolbox.topology_toolbox.KGTopologyToolbox.edge_degree_cardinality_summary) method:

```python
edge_dcs = kgtt.edge_degree_cardinality_summary()
```

> This will return a DataFrame with the cardinality patterns of the KG, where again values have been computed for each edge contained with the graph. 

Reviewer: ```suggestion
This will return a DataFrame with the cardinality patterns of the KG, where again values have been computed for each edge contained within the graph. 
```

> The values computed by the [`edge_degree_cardinality_summary`](https://graphcore-research.github.io/kg-topology-toolbox/generated/kg_topology_toolbox.topology_toolbox.KGTopologyToolbox.html#kg_topology_toolbox.topology_toolbox.KGTopologyToolbox.edge_degree_cardinality_summary) method include: one-to-one, one-to-many, many-to-one, many-to-many and other cardinality metrics.

Reviewer: ```suggestion
The values computed by the [`edge_degree_cardinality_summary`](https://graphcore-research.github.io/kg-topology-toolbox/generated/kg_topology_toolbox.topology_toolbox.KGTopologyToolbox.html#kg_topology_toolbox.topology_toolbox.KGTopologyToolbox.edge_degree_cardinality_summary) method include triple cardinality (one-to-one, one-to-many, many-to-one, many-to-many), head and tail degrees and other cardinality metrics.
```

# KG Topology Toolbox
![Continuous integration](https://github.com/graphcore-research/kg-topology-toolbox/actions/workflows/ci.yaml/badge.svg)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

![Python](https://img.shields.io/badge/python-3670A0?style=for-the-badge&logo=python&logoColor=ffdd54)
![NumPy](https://img.shields.io/badge/numpy-%23013243.svg?style=for-the-badge&logo=numpy&logoColor=white)
![SciPy](https://img.shields.io/badge/SciPy-%230C55A5.svg?style=for-the-badge&logo=scipy&logoColor=%white)
![Pandas](https://img.shields.io/badge/pandas-%23150458.svg?style=for-the-badge&logo=pandas&logoColor=white)

`kg-topology-toolbox` is a Python-based toolbox for computing topological properties of Knowledge Graphs (KGs). This library provides researchers and practitioners with tools to better understand the structural characteristics of KGs and how they might impact the predictive performance of any models trained upon them.

> `kg-topology-toolbox` enables computation of a variety of properties of KGs, focusing on **edge topological (Symmetric, Inverse, Inference, Loops & Composition) and cardinality patterns (one-to-one, one-to-many, many-to-one & many-to-many).** 

Reviewer: ```suggestion
`kg-topology-toolbox` enables computation of a variety of properties of KGs, focusing on **edge topological (symmetry, inverse, inference, loop and composition) and cardinality patterns (one-to-one, one-to-many, many-to-one & many-to-many).** 
```

Unlike other libraries, `kg-topology-toolbox` provides the ability to compute these properties at the level of individual triples, as well as at the level of relations.

> ![edge patterns](relational_patterns.png "Edge Patterns")

Reviewer: Maybe I would slightly reduce the size of the figure. Also, we can include loops (using the same figure that appears in the jupyter notebook, or similar)

Author: Good point - I moved to the notebook figure in 3913f3c

---

Full documentation can be found at https://graphcore-research.github.io/kg-topology-toolbox/

> For a walkthrough of the main functionalities of `kg-topology-toolbox`, we provide an introductory [Jupyter notebook](docs/source/notebooks/ogb_biokg_demo.ipynb). 

## Installation

The library has been tested on Ubuntu 20.04 & MacOS >= 14 and has been developed targetting Python >=3.9 - however it should be widely compatible with other systems.

To install the latest version of `kg-topology-toolbox` library, run:

```
pip install wheel
pip install git+https://github.com/graphcore-research/kg-topology-toolbox.git
```

If you would like to be able to change the source code and have the changes reflected in your environment, you can clone the repository and install the package in editable mode run:

```
git clone https://github.com/graphcore-research/kg-topology-toolbox.git
cd kg-topology-toolbox
pip install -e .
```

## Usage

Once installed, the library can be imported as follows:

```python
from kg_topology_toolbox import KGTopologyToolbox
```

> `kg-topology-toolbox` requires that the input KG is in the form of a pandas DataFrame with columns `h`, `r` and `t`. The `h` and `t` columns should contain the head and tail entities involved in the triple, and the `r` column should contain the relation type. These columns should be the integer identifiers of the entities and relations in the KG.

Reviewer: In principle, we don't strictly require the dataframe columns for be called h/r/t (see `*_column` optional parameters in [class constructor](https://graphcore-research.github.io/kg-topology-toolbox/generated/kg_topology_toolbox.topology_toolbox.KGTopologyToolbox.html#kg_topology_toolbox.topology_toolbox.KGTopologyToolbox)))... but it might be easier to say it like this.

Author: Thanks for pointing this out, I tried to clarify a bit in #555b223

For example, we can load a KG from a CSV file:

```python
import pandas as pd
df = pd.read_csv("path/to/kg.csv", columns=["h", "r", "t"])
```

This can then be used to instantiate a `KGTopologyToolbox` object:

```python
kgtt = KGTopologyToolbox(df)
```

### Computing Edge Topological Patterns

The `KGTopologyToolbox` object can be used to compute the topological properties of the KG. For example, to compute the edge patterns of the KG, we can use the `edge_pattern_summary` method:

```python
edge_eps = kgtt.edge_pattern_summary()
```

> This will return a DataFrame with the edge patterns of the KG, where values have been computed for each edge contained with the graph. The values contained in the dataframe include: 

Reviewer: ```suggestion
This will return a DataFrame with the edge patterns of the KG, where values have been computed for each edge contained with the graph. The values contained in the DataFrame include: 
```

>   - **is_loop**: True if the triple is a loop (``h == t``).

Reviewer: Not sure if I'm convinced by listing all the columns (here and below), it might make the README a bit too verbose - maybe referencing the notebook is enough? Not a strong opinion, though.

Author: Hey @AlCatt91 I do agree it is a bit verbose, but I thought it was important in the readme to try and convey what properties are possible to compute using this package. 

I'll think if there is a less verbose way to do this!

Author: I had a go at cleaning the text in c9a9b1fd

  - **n_inverse_relations**: The number of inverse relations r'.
  - **inverse_edge_types**: All relations r' (including r if the edge is symmetric) such that (t, r', h) is in the graph.
  - **has_inference**: True if the graph contains one or more triples (h, r', t) with ``r' != r``.
  - **n_inference_relations**: The number of inference relations r'.
  - **inference_edge_types**: All relations r' (including r) such that (h, r', t) is in the graph.
  - **has_composition**: True if the graph contains one or more triangles supported on the edge: (h, r1, x) + (x, r2, t).
  - **n_triangles**: The number of triangles.
  - **has_undirected_composition**: True if the graph contains one or more undirected triangles supported on the edge.
  - **n_undirected_triangles**: The number of undirected triangles (considering all edges as bidirectional).
  - **metapath_list**: The list of unique metapaths "r1-r2" for the directed triangles.

### Computing Edge Cardinality Patterns

Similarly, to compute the cardinality patterns of the KG, we can use the `edge_degree_cardinality_summary` method:

```python
edge_dcs = kgtt.edge_degree_cardinality_summary()
```

> This will return a DataFrame with the cardinality patterns of the KG, where again values have been computed for each edge contained with the graph. The dataframe contains the following columns:

Reviewer: ```suggestion
This will return a DataFrame with the cardinality patterns of the KG, where again values have been computed for each edge contained with the graph. The DataFrame contains the following columns:
```

  - **h_unique_rel**: Number of distinct relation types among edges with head entity h.
  - **h_degree**: Number of triples with head entity h.
  - **h_degree_same_rel**: Number of triples with head entity h and relation type r.
  - **t_unique_rel**: Number of distinct relation types among edges with tail entity t.
  - **t_degree**: Number of triples with tail entity t.
  - **t_degree_same_rel**: Number of triples with tail entity t and relation type r.
  - **tot_degree**: Number of triples with head entity h or tail entity t.
  - **tot_degree_same_rel**: Number of triples with head entity h or tail entity t, and relation type r.
  - **triple_cardinality**: cardinality type of the edge.

>   - **triple_cardinality_same_rel** (int): cardinality type of the edge in the subgraph of edges with relation type r.

Reviewer: ```suggestion
  - **triple_cardinality_same_rel**: cardinality type of the edge in the subgraph of edges with relation type r.
```

### Aggregating by Relation

It is also possible to aggregate the properties at the level of relations, you can use the `aggregate_by_relation` method:

```python
from kg_topology_toolbox.utils import aggregate_by_relation

relation_eps = aggregate_by_relation(edge_eps)
relation_dcs = aggregate_by_relation(edge_dcs)
```

> This will return a DataFrame with the edge patterns of the KG, where values have been aggregated for each unique relation contained within the graph.

Reviewer: ```suggestion
This will return a DataFrame with statistics for the edge properties, aggregated across edges of the same relation type, for all relations contained within the graph.
```
